ckan/lib/webassets_tools.py

```diff
@@ -1,172 +1,244 @@
-# encoding: utf-8
 from __future__ import annotations
 
 import logging
 import os
 import tempfile
 from typing import Any, Optional
+from typing_extensions import Literal, TypedDict, assert_never
 
-import yaml
 from markupsafe import Markup
 from webassets import Environment
 from webassets.loaders import YAMLLoader
 
 from ckan.common import config, g
 
 
-logger = logging.getLogger(__name__)
+log = logging.getLogger(__name__)
 env: Optional[Environment] = None
 
-yaml.warnings({u'YAMLLoadWarning': False})
+AssetType = Literal["style", "script"]
+
+
+class AssetCollection(TypedDict):
+    script: list[str]
+    style: list[str]
+    included: set[str]
 
 
 def create_library(name: str, path: str) -> None:
     """Create WebAssets library(set of Bundles).
     """
-    config_path = os.path.join(path, u'webassets.yml')
+    if not env:
+        log.debug("Webassets environment is not initialized yet")
+        return
+
+    config_path = os.path.join(path, "webassets.yaml")
+    if not os.path.exists(config_path):
+        config_path = os.path.join(path, "webassets.yml")
+
     if not os.path.exists(config_path):
+        log.debug(
+            "Cannot create library %s at %s because webassets.yaml is missing",
+            name,
+            path,
+        )
         return
-    assert env
+
     library: dict[str, Any] = YAMLLoader(config_path).load_bundles()
     bundles = {
-        u'/'.join([name, key]): bundle
+        f"{name}/{key}": bundle
         for key, bundle
         in library.items()
     }
 
-    # Unfortunately, you'll get an error attempting to register bundle
-    # with the same name twice. For now, let's just pop existing
-    # bundle and avoid name-conflicts
-    # TODO: make PR into webassets with preferable solution
-    # Issue: https://github.com/miracle2k/webassets/issues/519
+    # skip attempts to register an asset if name is taken. It gives us
+    # templates-like behavior, where the item that was registered first has
+    # highest priority.
     for name, bundle in bundles.items():
-        env._named_bundles.pop(name, None)
+        if is_registered(name):
+            log.debug(
+                "Skip registration of %s because it already exists",
+                name,
+            )
+            continue
+
+        # use absolute path to files. Otherwise they'll behave like templates
+        # and if plugin A and plugin B has their own `x.css`(i.e
+        # `ckanext/A/assets/x.css` and `ckanext/B/assets/x.css`), both plugins
+        # will load the same `x.css` from the plugin that was first registered
+        bundle.contents = [
+            os.path.join(path, item)
+            for item in bundle.contents
+        ]
+        log.debug("Register asset %s", name)
```

Comment:
I'm finding this comment a little bit confusing. On the previous if we write: `It gives us templates-like behaviour` but in this new comment it says: `Otherwise the'll behave like templates` so I'm a little bit confusing on what's going on here with the absolute paths.

Reply:
These are different parts of webassets.

 The previous comment describes named assets. One can register `xxx/style` and override existing `xxx/style` asset as we do in templates. That's exactly what we want.

Here we are talking about actual files. We don't want to override a single css/js file just by creating the file with the same name, as we do int templates. If some extension has `style.css`, I don't want this extension to use a different `style.css` that I created inside my extension.